src/renderer.rs

```diff
@@ -412,39 +412,53 @@ impl TextureBuffer {
     }
 
     /// Write an RGBA buffer at the specified location.
-    pub fn write_rgba_at(&mut self, buffer: &[u8], width: usize, pos: (usize, usize)) {
-        for row in 0..buffer.len() / width {
-            let dst_start = (pos.1 + row) * self.width + pos.0 * 4;
-            if dst_start >= self.inner.len() {
-                break;
-            }
-            let dst_row = &mut self.inner[dst_start..];
-
-            let src_start = row * width;
-            let src_row = &buffer[src_start..src_start + cmp::min(width, dst_row.len())];
-
-            let pixels = src_row.chunks(4).enumerate().filter(|(_i, pixel)| pixel != &[0, 0, 0, 0]);
-            for (i, pixel) in pixels {
-                dst_row[i * 4..i * 4 + 4].copy_from_slice(pixel)
-            }
-        }
+    pub fn write_rgba_at(&mut self, buffer: &[u8], width: usize, pos: (isize, isize)) {
+        self.write_buffer_inner::<4>(buffer, width, pos)
     }
 
     /// Write an RGB buffer at the specified location.
-    pub fn write_rgb_at(&mut self, buffer: &[u8], width: usize, pos: (usize, usize)) {
+    pub fn write_rgb_at(&mut self, buffer: &[u8], width: usize, pos: (isize, isize)) {
+        self.write_buffer_inner::<3>(buffer, width, pos);
+    }
+
+    fn write_buffer_inner<const N: usize>(
+        &mut self,
+        buffer: &[u8],
+        mut width: usize,
+        pos: (isize, isize),
+    ) {
+        // Clamp dst to zero.
+        let dst_x = pos.0.max(0);
+
+        // Compute the amount to cut in the src buffer.
+        let to_cut_x = (dst_x - pos.0).abs() as usize;
+        width -= to_cut_x * N;
+
         for row in 0..buffer.len() / width {
-            let dst_start = (pos.1 + row) * self.width + pos.0 * 4;
-            if dst_start >= self.inner.len() {
+            // Apply `y`.
+            let mut dst_start = (pos.1 + row as isize) * self.width as isize;
+
+            // Cut top.
+            if dst_start < 0 {
+                continue;
+            }
+
+            // Apply `x`.
+            dst_start += dst_x * N as isize;
```

Comment:
Can probably combine this with the original `dst_start` again?

For the first row it shouldn't matter if `< 0` is because the row or the column is negative.

Reply:
Couldn't it make `y` part positive then if ``dst_x * N`.abs() > dst_start.abs()`?

Reply:
Yeah but to be fair we don't do any checks on the "right" boundary to prevent bleeding into the next line. At this point it's only on the left so I thought this would be "consistent".

Ideally you'd also clamp on the right side, so it doesn't bleed into the next line.